### PR TITLE
Fixes #25584 - Improve help for compute resources

### PR DESCRIPTION
**Pull request by ofedoren**
The issue actually is about showing unclear errors, but since the error (e.g. mentioned in the issue) comes from the server, I'd say it is a server side issue.

What I'm proposing is to improve help section for CR in host creation help. Particularly to emphasize which attributes are required (e.g. `path` mentioned in the issue). Unfortunatelly I've got no environment (except libvirt and a dummy created from VMware) to test which attributes are really required. The chages are based on UI and [CR models](https://github.com/theforeman/foreman/tree/develop/app/models/compute_resources/foreman/model) (basically on `default_vm_attributes`/`new_vm`/`create_vm`, etc)

Requires https://github.com/theforeman/hammer-cli/pull/309 to be merged first. Also this PR is built above https://github.com/theforeman/hammer-cli-foreman/pull/424, since there are chages which will lead to merge conflicts.

### CODE REVIEW

**Comment by ofedoren** [comment]
@shiramax, rebased.

**Comment by shiramax** [comment]
thanks @ofedoren 
We will need to confirm that these are the right attributes for each compute resource 
I will verify that for oVirt,
@kgaikwad - can you please confirm that for GCE and EC2?
@ares - can you please confirm that for OpenStack?
@ezr-ondrej - can you please confirm that for VMware?

File: lib/hammer_cli_foreman/compute_resource/openstack.rb [hunk 1]
```diff
@@ -6,7 +6,7 @@ def name
       end
 
       def compute_attributes
-        %w[flavor_ref image_ref tenant_id security_groups network]
+        %w[availability_zone boot_from_volume flavor_ref image_ref tenant_id scheduler_hint_filter security_groups network]
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/openstack.rb]: same here

File: lib/hammer_cli_foreman/compute_resource/gce.rb [hunk 2]
```diff
@@ -6,7 +6,7 @@ def name
       end
 
       def compute_attributes
-        %w[machine_type image_id network external_ip]
+        %w[associate_external_ip image_id machine_type network user_data]
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/gce.rb]: same here

File: lib/hammer_cli_foreman/compute_resource/ec2.rb [hunk 3]
```diff
@@ -6,7 +6,7 @@ def name
       end
 
       def compute_attributes
-        %w[flavor_id image_id availability_zone security_group_ids managed_ip]
+        %w[availability_zone flavor_id groups image_id security_group_ids managed_ip]
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/ec2.rb]: can we find the description for each attribute?

File: lib/hammer_cli_foreman/compute_resource/ovirt.rb [hunk 4]
```diff
@@ -10,7 +10,10 @@ def compute_attributes
           ['cluster',  _('ID of cluster to use')],
           ['template', _('Hardware profile to use')],
           ['cores',    _('Integer value, number of cores')],
-          ['memory',   _('Amount of memory, integer value in bytes')]
+          ['sockets',  _('Integer value, number of sockets')],
+          ['memory',   _('Amount of memory, integer value in bytes')],
+          ['image_id'],
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/ovirt.rb]: image_id is a general attribute, shouldn't be here.

File: lib/hammer_cli_foreman/compute_resource/ovirt.rb [hunk 5]
```diff
@@ -10,7 +10,10 @@ def compute_attributes
...
+          ['user_data', _('Comment')]
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/ovirt.rb]: I'm not sure that's spouse to be here... how did you find it?

**Comment by ares** [comment]
redirecting to @lzap - can you please confirm that for OpenStack? I guess it would need to be at https://github.com/fog/fog-openstack/blob/master/lib/fog/openstack/compute/models/server.rb but I don't see boot_from_volume and scheduler_hint_filter

but if @lzap has some openstack instnace, it's easier to test on existing vm

File: lib/hammer_cli_foreman/compute_resource/gce.rb (hunk 2 above)
Comment (kgaikwad) [inline lib/hammer_cli_foreman/compute_resource/gce.rb]: Thank you @ofedoren for updating attributes here.  
Here, `user_data` attribute is not required as while creating an image, user can have this attribute.

File: lib/hammer_cli_foreman/compute_resource/vmware.rb [hunk 6]
```diff
@@ -9,13 +9,13 @@ def name
 
       def compute_attributes
         [
-          ['cpus',                 _('CPU count')],
-          ['corespersocket',       _('Number of cores per socket (applicable to hardware versions < 10 only)')],
-          ['memory_mb',            _('Integer number, amount of memory in MB')],
+          ['cpus',                 _('CPU count'), { bold: true }],
+          ['corespersocket',       _('Number of cores per socket (applicable to hardware versions < 10 only)'), { bold: true }],
+          ['memory_mb',            _('Integer number, amount of memory in MB'), { bold: true }],
           ['firmware',             'automatic/bios/efi'],
```
Comment (ezr-ondrej) [inline lib/hammer_cli_foreman/compute_resource/vmware.rb]: Could we put the `firmware` after the required attributes? Now it feels bit weird to be here.
Or sort it alphabeticaly, but now it feels bit messy ordered :)

**Comment by ezr-ondrej** [comment]
Could we somehow add a default values for the not required attributes, which have defaults? I do not insist doing it in this PR, though.

**Comment by ofedoren** [comment]
@shiramax, @kgaikwad, @ezr-ondrej, thanks! The required attributes are highlighted in bold in the help (the note at the beggining of the `provider specific atrributes` section tells it), the others are plain text.

But I agree that putting the required attributes at the top of the list is a good idea. :)

@ezr-ondrej, yes, I guess we can add a default values for the not required attributes, but I assume it is done on the server side by `normalize...` methods or `default_attrs..`.

**Comment by ezr-ondrej** [comment]
> @ezr-ondrej, yes, I guess we can add a default values for the not required attributes, but I assume it is done on the server side by `normalize...` methods or `default_attrs..`.

Yeah it is, but there is no way for the `hammer` user to know what those values are going to be. That is my point, to add them in the help :) But I agree it would be better to get them somehow from the `apipie`, not to hardcode them.

**Comment by lzap** [comment]
I am giving the credentials to the RHOS I test with, all employees can use it.

**Comment by ofedoren** [comment]
@shiramax, updated.

File: lib/hammer_cli_foreman/compute_resource/libvirt.rb [hunk 7]
```diff
@@ -7,29 +7,32 @@ def name
...
+          ['start', _('Boolean (expressed as 0 or 1), whether to start the machine or not')]
         ]
       end
 
       def interface_attributes
         [
           ['type',   _('Possible values: %s') % 'bridge, network'],
           ['bridge', _('Name of interface according to type')],
-          ['model',  _('Possible values: %s') % 'virtio, rtl8139, ne2k_pci, pcnet, e1000']
+          ['model',  _('Possible values: %s') % 'virtio, rtl8139, ne2k_pci, pcnet, e1000'],
```
Comment (shiramax) [inline lib/hammer_cli_foreman/compute_resource/libvirt.rb]: In libevirt we have these three options: 
image
i'm not sure about the bridge, do you know why it's not in the UI is well?

Comment (ofedoren) [reply lib/hammer_cli_foreman/compute_resource/libvirt.rb]: It is, when you're creating a host via libvirt and go Interfaces -> New/Edit -> Select `Type`.

Also, see https://github.com/theforeman/foreman/blob/develop/app/models/compute_resources/foreman/model/libvirt.rb#L224-L228. I guess the API does support it.

**Comment by ares** [comment]
On August 14, 2019 16:26:58 Shira Maximov <notifications@github.com> wrote:

> shiramax requested changes on this pull request.
>
>
>
>>          ]
>       end
>
>       def interface_attributes
>         [
>           ['type',   _('Possible values: %s') % 'bridge, network'],
>           ['bridge', _('Name of interface according to type')],
> -          ['model',  _('Possible values: %s') % 'virtio, rtl8139, 
> ne2k_pci, pcnet, e1000']
> +          ['model',  _('Possible values: %s') % 'virtio, rtl8139, 
> ne2k_pci, pcnet, e1000'],
>
> In libevirt we have these three options:
> image
> i'm not sure about the bridge, do you know why it's not in the UI is well?
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/hammer-cli-foreman/pull/427#pullrequestreview-274923612

I suppose the bridge in UI only shows if you change network type. In 
API/hammer we can't hide fields that depends on other fields values.

--
Marek

File: doc/host_create.md [hunk 8]
```diff
@@ -263,13 +268,14 @@ cluster               Cluster ID from VMware
 resource_pool         Resource Pool ID from VMware
 path                  Path to folder
 guest_id              Guest OS ID form VMware
-scsi_controller_type  ID of the controller from VMware
 hardware_version      Hardware version ID from VMware
 add_cdrom             Must be a 1 or 0, Add a CD-ROM drive to the virtual machine
 cpuHotAddEnabled      Must be a 1 or 0, lets you add memory resources while the machine is on
-memoryHotAddEnabled   Must be a 1 or 0, lets you add CPU resources while the machine is on
-start                 Must be a 1 or 0, whether to start the machine or not
```
Comment (shiramax) [inline doc/host_create.md]: why did you remove start from here? it should still be here in when creating a host.

File: doc/host_create.md [hunk 9]
```diff
@@ -195,26 +194,29 @@ Available keys for `--compute-attributes`:
 ```
 cpus          # number of CPUs
 memory        # string, amount of memory, value in bytes
-start         # Must be a 1 or 0, whether to start the machine or not
```
Comment (shiramax) [inline doc/host_create.md]: why did you remove start from here? it should still be here in when creating a host.

**Comment by ofedoren** [comment]
@shiramax, thanks! Updated.

Also, just reminding that https://github.com/theforeman/hammer-cli/pull/309 is required to be merged first. Althought tests are passing, there can be a runtime error.

**Comment by shiramax** [comment]
@ofedoren can you please rebase?

**Comment by ofedoren** [comment]
@shiramax, rebased.

File: doc/host_create.md [hunk 10]
```diff
@@ -195,26 +194,30 @@ Available keys for `--compute-attributes`:
...
 ```
 
 Available keys for `--interface`:
 ```
-compute_type                      # one of [bridge, network]
-compute_network / compute_bridge  # name of interface according to type
-compute_model                     # one of [virtio, rtl8139, ne2k_pci, pcnet, e1000]
+type              # one of [bridge, network]
+network / bridge  # name of interface according to type
+model             # one of [virtio, rtl8139, ne2k_pci, pcnet, e1000]
```
Comment (shiramax) [inline doc/host_create.md]: after the rebase, all the --interface attributes,  it should stay with "compute_" prefix.

File: doc/host_create.md [hunk 11]
```diff
@@ -228,14 +231,16 @@ Available keys for `--compute-attributes`:
...
 start      # Must be a 1 or 0, whether to start the machine or not
 ```
 
 Available keys for `--interface`:
 ```
-compute_name         # eg. eth0
-compute_network      # select one of available networks for a cluster
+name         # eg. eth0
+network      # select one of available networks for a cluster
+interface    # interface type
```
Comment (shiramax) [inline doc/host_create.md]: after the rebase, all the --interface attributes,  it should stay with "compute_" prefix.

**Comment by ofedoren** [comment]
@shiramax, updated. Thanks, I forgot to change those as well :(

**Comment by kgaikwad** [comment]
Apologize for late post.
In case of GCE/Ec2, user can configure `image_id` inside compute_attributes at compute_profile as well as host level. But removing `image_id` from compute_attributes in gce.rb broken the flow.

Am I missing anything here? Is there any foreman PR which handles this?

**Comment by ofedoren** [comment]
@kgaikwad, apologize for late reply.. Thanks for tracking that though :)

In `hammer host create` command I think we handle it though `--image-id` (see https://github.com/theforeman/hammer-cli-foreman/blob/9c4a2c7cffdb7255b1703986e29360a972be17da/lib/hammer_cli_foreman/hosts/common_update_options.rb#L102).

But in `hammer compute-profile values create/update` command I don't see any mentions how to set an image... 

@shiramax, @kgaikwad what do you think will be better resolution
  - Add `--image-id` option as we do have in `hammer host create` command or
  - Return `image_id` into `compute_attributes` help section?

**Comment by kgaikwad** [comment]
@ofedoren,

There was no separate option in the compute profile to set an image. 
Previously it was done by using `image_id` value inside `--compute-attributes` option.
Example -
`hammer  compute-profile values update  --compute-attributes "image_id=3780108136525169178,machine_type=f1-micro,associate_external_ip=true,network=default" --compute-profile-id 4 --compute-resource-id 6 --volume "size_gb=20"`

And with removal of `image_id` from help section, it will be difficult for new user to know about how to set image_id for the compute_profile. 


As `--image-id` option present in `host` commands and to maintain a consistency as well as to avoid confusion between host & compute_profile commands usage,  it would be good to have similar option for compute_profile.

@shiramax, what do you think?

**Comment by shiramax** [comment]
@kgaikwad image should not be inside compute profile at all. 
Compte profile should describe the hardware configuration of the host - for example : cores, sockets, memory, interfaces, and volumes.
the OS is not relevant for compute profile.
also, when you create a host from an image, you need to specify Architecture and Operating system , only then you will be able to choose an Image,  Architecture and Operating system  should not be inside compute profile.

**Comment by kgaikwad** [comment]
@shiramax,
As far as I know we have an options on web UI to set architecture, operating system and image at compute profile level.
If user want to set those values at host level, he can.

Could you please confirm if there is any recent change into compute profile form?

**Comment by kgaikwad** [comment]
@shiramax,
Now, I got it. That's why may be if user don't select OS and architecture values first before selecting the compute profile, image field from OS tab on host form gets disabled. To enable that field user needs to deselect compute profile.

**Comment by shiramax** [comment]
exactly,
AFIK images should not be inside the compute profile.

**Comment by kgaikwad** [comment]
@shiramax, 

As mentioned web UI is inconsistent with what we are saying. My concerns are as follows:
1. On web UI compute profile page shows an image field for selection and on hammer side it is not showing in help and don't have any option to configure image value.
2. There is no place except the compute profile page to configure an image value for hosts. That means every time user needs to select an image value on Host form.

We will need to either consider redesigning those web UI pages or provide an option which will allow user to configure an image.

**Comment by shiramax** [comment]
@kgaikwad 
Adding @ShimShtein and @apuntamb  since we spoke about this issue last week. I agree that the current design is not ideal, and we should re-think about it. maybe you can open a topic in https://community.theforeman.org ?
